swap/tests/alice_and_bob_refund_using_cancel_and_refund_command_timelock_not_expired.rs

```diff
@@ -0,0 +1,108 @@
+pub mod harness;
+
+use harness::alice_run_until::is_xmr_lock_transaction_sent;
+use harness::bob_run_until::is_btc_locked;
+use harness::SlowCancelConfig;
+use swap::protocol::alice::AliceState;
+use swap::protocol::bob::BobState;
+use swap::protocol::{alice, bob};
+
+#[tokio::test]
+async fn given_alice_and_bob_manually_cancel_when_timelock_not_expired_errors() {
+    harness::setup_test(SlowCancelConfig, |mut ctx| async move {
+        let (bob_swap, bob_join_handle) = ctx.bob_swap().await;
+        let swap_id = bob_swap.id;
+        let bob_swap = tokio::spawn(bob::run_until(bob_swap, is_btc_locked));
+
+        let alice_swap = ctx.alice_next_swap().await;
+        let alice_swap = tokio::spawn(alice::run_until(alice_swap, is_xmr_lock_transaction_sent));
+
+        let bob_state = bob_swap.await??;
+        assert!(matches!(bob_state, BobState::BtcLocked { .. }));
+
+        let (bob_swap, bob_join_handle) = ctx
+            .stop_and_resume_bob_from_db(bob_join_handle, swap_id)
+            .await;
+        assert!(matches!(bob_swap.state, BobState::BtcLocked { .. }));
+
+        let alice_state = alice_swap.await??;
+        assert!(matches!(
+            alice_state,
+            AliceState::XmrLockTransactionSent { .. }
+        ));
+
+        // Bob tries but fails to manually cancel
+        let result = bob::cancel(bob_swap.id, bob_swap.bitcoin_wallet, bob_swap.db, false)
+            .await?
+            .unwrap_err();
+        assert!(matches!(
+            result,
+            bob::cancel::Error::CancelTimelockNotExpiredYet
+        ));
+
+        ctx.restart_alice().await;
+        let alice_swap = ctx.alice_next_swap().await;
+        assert!(matches!(
+            alice_swap.state,
+            AliceState::XmrLockTransactionSent { .. }
+        ));
+
+        // Alice tries but fails manual cancel
+        let result = alice::cancel(
+            alice_swap.swap_id,
+            alice_swap.bitcoin_wallet,
+            alice_swap.db,
+            false,
+        )
+        .await?
+        .unwrap_err();
+        assert!(matches!(
+            result,
+            alice::cancel::Error::CancelTimelockNotExpiredYet
+        ));
+
+        let (bob_swap, bob_join_handle) = ctx
+            .stop_and_resume_bob_from_db(bob_join_handle, swap_id)
+            .await;
+        assert!(matches!(bob_swap.state, BobState::BtcLocked { .. }));
+
+        // Bob tries but fails to manually refund
+        bob::refund(bob_swap.id, bob_swap.bitcoin_wallet, bob_swap.db, false)
+            .await?
+            .unwrap_err();
+
+        let (bob_swap, _) = ctx
+            .stop_and_resume_bob_from_db(bob_join_handle, swap_id)
+            .await;
+        assert!(matches!(bob_swap.state, BobState::BtcLocked { .. }));
+
+        ctx.restart_alice().await;
+        let alice_swap = ctx.alice_next_swap().await;
+        assert!(matches!(
+            alice_swap.state,
+            AliceState::XmrLockTransactionSent { .. }
+        ));
```

Comment:
Same as above, I think we should get rid of these "pre-check" asserts.

Ideally, a test only contains an assert for things that matter to the test which in this case is about timelocks not being expired yet.

Reply:
I disagree :) 
This statement makes sure that nothing along the way of restarting the application is causing a side effect that changes the state. 
Granted, we only simulate the restart, but still, I think it is good that we ensure that we are the expected state after the restart.
In these rather complex e2e tests it is good to test all the assumptions on the way - if we don't do that it will be even harder to debug once a test fails, because we only fail in the end and don't know why!

Reply:
Let's agree to disagree then.

I just know from my own experience that tests which are noisy are seldomly checked for correctness and what is the point of a test if I don't verify that it is even correct?

>  In these rather complex e2e tests it is good to test all the assumptions on the way - if we don't do that it will be even harder to debug once a test fails, because we only fail in the end and don't know why!

That sounds more like a case for better logs than littering the test with assertions. Can also always set a breakpoint and inspect the state :man_shrugging:

Reply:
I feel this comment is completely independent of work done in this PR but does concern our tests in general. Let's clean this up in a follow up if we deem it important.

Reply:
Agreed, I just pointed it out because these tests make excessive use of these asserts :)